### ISSUE TITLE
Vision for MkDocs on Read the Docs

### ISSUE THREAD

**Issue by davidfischer** [readthedocs/readthedocs.org#4588]
There was quiet a bit of discussion in #4556 about MkDocs on Read the Docs. I wanted to move some of that discussion here as that is an active PR that I am hoping to merge soonish.

I don't want to speak for people other than myself as the RTD maintainer who has worked on MkDocs more recently than anyone else at RTD.

## Where RTD would like to go with MkDocs

* We have an [MkDocs Milestone](https://github.com/rtfd/readthedocs.org/milestone/24) although much of it is bugfixes
* More documentation and a full "getting started with MkDocs on RTD" guide as well as a "customizing MkDocs" section
* Additional integrations with MKDocs via [plugins](https://www.mkdocs.org/user-guide/configuration/#plugins). This part may be more of a community effort.

## Things RTD would like from MkDocs

* **Search Indexing** RTD was using the `json` builder ([removed in 0.17](https://www.mkdocs.org/about/release-notes/#other-changes-and-additions-to-version-0170), see #4205) in order to index for searching. It would be nice to have this or something like it as opposed to crawling/indexing the HTML.
* **Format stability** Stability in the `mkdocs.yml` format especially around themes. There were quite a few backward incompatible changes here from 0.15.x to 1.x. Because RTD supports multiple versions of MkDocs, this was an issue. This may be less of an issue once we stop overriding the theme.

## Things MkDocs needs from RTD

I'm leaving this section a bit bare for now as I don't want to speak for the MkDocs maintainers. Some things mentioned in #4556 included:

* **Honoring users wishes for themes and plugins.** RTD supports plugins entirely although the process could be clearer. RTD had support for themes although bugs were introduced with the changes to `mkdocs.yml` in 0.17. After #4556, it should support different themes better.

**Comment by waylan** [comment]
## Things MkDocs needs from RTD

In addition to the item listed above, we also need:

* **Clear documentation of what RTD depends on**. Ideally, there would be one location which would list all of the hooks/APIs that RTD uses. For example, if such a list existed, we might not have removed the `json` builder, as we were of the impression that it was unused. Yes, this information is available by reading the code, but that's not practical every time we need to make a quick check. Note that the requested documentation is not for end users, but for MkDocs developers, Theme developers, and possibly plugin developers. It might not fit into your public documentation, so a wiki page or the like would be fine.

**Comment by stsewd** [comment]
@waylan thanks, sorry for the so late reply. We are now focusing on improving the MkDocs experience on RTD, we are going to be indexing search data from the `search/search_index.json` file. So it would be nice if any change to the structure of that file would be kipped or versioned if there are changes. I'll be writing the document you mention in the next days/weeks.
